#!/usr/bin/env cwl-runner

cwlVersion: v1.0
class: CommandLineTool
baseCommand: perl
label: Removes reads present in the input list of readnames from the input file of reads.

hints:
  DockerRequirement:
    dockerPull: ttubb/1d2scripts:release-0.2.0
  SoftwareRequirement:
    #Requires script: https://raw.githubusercontent.com/gringer/bioinfscripts/master/fastx-fetch.pl
    packages:
      perl:
        specs: [ https://perldoc.perl.org/5.28.1/perl.html ]
        version: [ "5.28.1" ]

arguments: [/software/fastx-fetch.pl, -v]

inputs:
  reads_1d:
    label: Reads generated by guppy_basecaller.
    type: File
    inputBinding:
      position: 4
  readnames_1d2:
    label: Sequencing summary generated by guppy_basecaller_1d2.
    type: File
    inputBinding:
      position: 3
      prefix: -i

outputs:
  reads_1d_filtered:
    label: reads_1d minus all reads present in readnames_1d2.
    type: stdout
    format: edam:format_1930 #fastq

stdout: $(inputs.reads_1d.basename)

s:author:
  - class: s:Person
    s:email: mailto:tom.tubb@googlemail.com
    s:name: Tom Tubbesing
s:dateCreated: "2019-04-16"
s:license: "https://spdx.org/licenses/GPL-3.0-or-later.html"

$namespaces:
  s: http://schema.org/
  edam: http://edamontology.org/

$schemas:
  - https://schema.org/version/latest/schema.rdf
  - http://edamontology.org/EDAM_1.18.owl
