#!/usr/bin/env cwl-runner

cwlVersion: v1.0
class: Workflow
label: 
doc: |
  Uses guppy to perform basecalling on raw nanopore read data from experiments using 1d2 chemistry. Returns report files and seperate files containing:
    -all 1D and 1D2 reads as basecalled by 1D basecaller
    -all 1D2 reads as basecalled by 1D2 basecaller
    -a combination of all 1D reads as basecalled by 1D basecaller and all 1D2 reads as basecalled by 1D2 basecaller.
    IMPORTANT: Either kit id and flowcell id OR custom config files OR config names have to be provided.

hints:
  SoftwareRequirement:
    packages:
      guppy:
        specs: [ https://community.nanoporetech.com/protocols/Guppy-protocol-preRev/v/gpb_2003_v1_revk_14dec2018/linux-guppy ] #Login credentials required.
        version: [ "3.3.0" ]
      perl:
        specs: [ https://perldoc.perl.org/5.28.1/perl.html ]
        version: [ "5.28.1" ]
      awk:
        specs: [ https://pubs.opengroup.org/onlinepubs/9699919799/utilities/awk.html ]
        version: [ "4.1.4" ]


requirements:
  SubworkflowFeatureRequirement: {}
  InlineJavascriptRequirement: {}
  StepInputExpressionRequirement: {}
  MultipleInputFeatureRequirement: {}

inputs:
  fast5_directory:
    label: Directory containing raw nanopore signal data in fast5 format.
    type: Directory
  worker_threads:
    label: Number of CPU-threads used in computationally intensive steps.
    type: int
    default: 1
  flowcell_id:
    label: Identifier of flowcell used in nanopore experiment (e.g. FLO-MIN106). 
    type: string?
  kit_id:
    label: Identifier of kit used in nanopore experiment (e.g. SQK-RBK004).
    type: string?
  config_file_1d:
    label: A custom config file to control guppy 1d parameters in detail.
    type: File?
  config_name_1d:
    label: Name of one of guppys preset configs for 1d basecalling.
    type: string?
  config_file_1d2:
    label: A custom config file to control guppy 1d2 parameters in detail.
    type: File?
  config_name_1d2:
    label: Name of one of guppys preset configs for 1d2 basecalling.
    type: string?   
  enable_adapter_trimming:
    label: Determines wether guppy will perform adapter trimming.
    type: string
    default: "false"

outputs:
  merged_reads:
    label: fastq file with all reads generated by basecaller.
    type: File
    format: edam:format_1930 #fastq
    outputSource: all_read_merging/merged_reads
  merged_1d_reads:
    label: fastq file with reads generated by 1d basecaller minus 1d2 reads.
    type: File
    format: edam:format_1930 #fastq
    outputSource: rename_1d_reads/file_renamed
  merged_1d2_reads:
    label: fastq file with reads generated by 1d2 basecaller.
    type: File
    format: edam:format_1930 #fastq
    outputSource: rename_1d2_reads/file_renamed
  sequencing_summary:
    label: Array containing summary of results reported by 1d- and 1d2-basecaller.
    type: File[]
    format: edam:format_1964 #plain text
    outputSource: [rename_1d_summary/file_renamed, rename_1d2_summary/file_renamed]

steps:
  1d_basecalling:
    label: Performs 1d basecalling on nanopore signal data using ONTs guppy software.
    run: guppyBasecallerTool.cwl
    in:
      fast5_directory: fast5_directory
      flowcell_id: flowcell_id
      kit_id: kit_id
      worker_threads: worker_threads
      enable_adapter_trimming: enable_adapter_trimming
      config_name: config_name_1d
      config_file: config_file_1d
    out: [array_of_reads, reads_directory, sequencing_summary, sequencing_telemetry]
  1d2_basecalling:
    label: Performs 1d2 basecalling on nanopore signal data using ONTs guppy software.
    run: guppy1d2BasecallerTool.cwl
    in:
      fast5_directory: fast5_directory
      index_file: 1d_basecalling/sequencing_summary
      worker_threads: worker_threads
      flowcell_id: flowcell_id
      kit_id: kit_id
      config_name: config_name_1d2
      config_file: config_file_1d2
      enable_adapter_trimming: enable_adapter_trimming
    out: [array_of_reads, reads_directory, sequencing_summary, sequencing_telemetry]      
  1d_read_merging:
    label: Merges all 1d reads into a single file.
    run: ../misc/readMergingTool.cwl
    in:
      array_of_reads: 1d_basecalling/array_of_reads
      namesource_directory: fast5_directory
    out: [merged_reads]
  1d2_read_merging:
    label: Merges all 1d2 reads into a single file.
    run: ../misc/readMergingTool.cwl
    in:
      array_of_reads: 1d2_basecalling/array_of_reads
      namesource_directory: fast5_directory
    out: [merged_reads]
  read_filtering:
    label: Removes all 1d2 reads from the file of 1d reads.
    run: filter1d2Workflow.cwl
    in:
      reads_1d: 1d_read_merging/merged_reads
      reads_1d2: 1d2_read_merging/merged_reads
    out: [reads_1d_filtered]
  rename_1d_reads:
    label: Renames file containing 1d reads.
    run: ../misc/renameTool.cwl
    in:
      namesource: fast5_directory
      rename_this: read_filtering/reads_1d_filtered
      newname:
        valueFrom: $((inputs.namesource.basename)+"-reads-1d.fastq")
    out: [file_renamed]
  rename_1d2_reads:
    label: Renames file containing 1d2 reads.
    run: ../misc/renameTool.cwl
    in:
      namesource: fast5_directory
      rename_this: 1d2_read_merging/merged_reads
      newname:
        valueFrom: $((inputs.namesource.basename)+"-reads-1d2.fastq")
    out: [file_renamed]
  rename_1d_summary:
    label: Renames sequencing summary of 1d basecaller to avoid identical names.
    run: ../misc/renameTool.cwl
    in:
      namesource: fast5_directory
      rename_this: 1d_basecalling/sequencing_summary
      newname:
        valueFrom: $((inputs.namesource.basename)+"-1d_basecalling_summary.txt")
    out: [file_renamed]
  rename_1d2_summary:
    label: Renames sequencing summary of 1d2 basecaller to avoid identical names.
    run: ../misc/renameTool.cwl
    in:
      namesource: fast5_directory
      rename_this: 1d2_basecalling/sequencing_summary
      newname:
        valueFrom: $((inputs.namesource.basename)+"-1d2_basecalling_summary.txt")
    out: [file_renamed]
  all_read_merging:
    label: Merges the 1d and 1d2 reads into a single file.
    run: ../misc/readMergingTool.cwl
    in:
      array_of_reads:
        source: [read_filtering/reads_1d_filtered, rename_1d2_reads/file_renamed]
        linkMerge: merge_flattened
      namesource_directory: fast5_directory
    out: [merged_reads]

s:author:
  - class: s:Person
    s:email: mailto:tom.tubb@googlemail.com
    s:name: Tom Tubbesing
s:dateCreated: "2019-04-16"
s:license: none

$namespaces:
  s: http://schema.org/
  edam: http://edamontology.org/

$schemas:
  - https://schema.org/version/latest/schema.rdf
  - http://edamontology.org/EDAM_1.18.owl
