#!/usr/bin/env cwl-runner

cwlVersion: v1.0
class: Workflow
label: Creates a hybrid assembly from long, raw ONT nanopore reads and short Illumina reads.
doc: |
  This workflow uses raw long reads from a 1D-squared-chemistry ONT nanopore sequencing experiment to create a draft assembly. The draft assembly is subsequently polished using Illumina short reads. Long reads from 1D- and 1D-squared-basecalling a pooled and used together after the basecalling step. The workflow also creates reports regarding the quality of nanopore- and Illumina-reads. It will generate a Quast report on the quality of the resulting assembly. If a reference genome is provided, Quast will try to align the assembly to the reference.

hints:
  cwltool:LoadListingRequirement:
    loadListing: shallow_listing
  SoftwareRequirement:
    packages:
      guppy:
        specs: [ "https://community.nanoporetech.com/protocols/Guppy-protocol-preRev/v/gpb_2003_v1_revk_14dec2018/linux-guppy" ] #Login credentials required.
        version: [ "3.1.5" ]
      perl:
        specs: [ "https://perldoc.perl.org/5.28.1/perl.html" ]
        version: [ "5.28.1" ]
      awk:
        specs: [ "https://pubs.opengroup.org/onlinepubs/9699919799/utilities/awk.html" ]
        version: [ "4.1.4" ]
      minimap2:
        specs: [ "https://github.com/lh3/minimap2" ]
        version: [ "2.16" ]
      miniasm:
        specs: [ https://biosphere.france-bioinformatique.fr/wikia2/index.php/Miniasm
                 https://github.com/lh3/miniasm ]
        version: [ "0.3" ]
      racon:
        specs: [ "https://github.com/isovic/racon" ]
        version: [ "1.3.2" ]
      bowtie2:
        specs: [ "https://github.com/BenLangmead/bowtie2" ]
        version: [ "2.3.5.1" ]
      pilon:
        specs: [ "https://identifiers.org/RRID:SCR_014731" ]
        version: [ "1.23" ]
      samtools:
        specs: [ "https://identifiers.org/RRID:SCR_002105" ]
        version: [ "1.7" ]
      fastqc:
        specs: [ "https://identifiers.org/RRID:SCR_014583" ]
        version: [ "0.11.5" ]
      nanoqc:
        specs: [ "https://github.com/wdecoster/nanoQC" ]
        version: [ "0.9.0" ]
      nanoplot:
        specs: [ "https://github.com/wdecoster/NanoPlot" ]
        version: [ "1.22.0" ]
      quast:
        specs: [ "https://identifiers.org/RRID:SCR_001228" ]
        version: [ "5.0.2" ]
        
requirements:
  SubworkflowFeatureRequirement: {}
  StepInputExpressionRequirement: {}
  ScatterFeatureRequirement: {}
  MultipleInputFeatureRequirement: {}

inputs:
  worker_threads:
    label: Number of CPU-threads used in computationally intensive steps.
    type: int
  fast5_directory:
    label: Directory containing raw ONT nanopore reads in fast5 format.
    type: Directory
  short_reads_fwd:
    label: Illumina forward reads.
    type: File
  short_reads_rev:
    label: Illumina reverse reads.
    type: File
  flowcell_id:
    label: Identifier of flowcell used in nanopore experiment (e.g. FLO-MIN106). Please provide either flowcell & kit IDs or config files/config file names. Do not provide both.
    type: string?
  kit_id:
    label: Identifier of kit used in nanopore experiment (e.g. SQK-RBK004). Please provide either flowcell & kit IDs or config files/config file names. Do not provide both.
    type: string?
  config_file_1d:
    label: A custom config file to control guppy 1d parameters in detail. Please provide either flowcell & kit IDs or config files/config file names. Do not provide both.
    type: File?
  config_name_1d:
    label: Name of one of guppys preset configs for 1d basecalling. Please provide either flowcell & kit IDs or config files/config file names. Do not provide both.
    type: string?
  config_file_1d2:
    label: A custom config file to control guppy 1d2 parameters in detail. Please provide either flowcell & kit IDs or config files/config file names. Do not provide both.
    type: File?
  config_name_1d2:
    label: Name of one of guppys preset configs for 1d2 basecalling. Please provide either flowcell & kit IDs or config files/config file names. Do not provide both.
    type: string?
  reference_genome:
    label: Reference genome that will be used for quality control by quast.
    type: File?
    
outputs:
  ont_reads:
    label: Basecalling ONT nanopore reads.
    type: File
    outputSource: basecalling/merged_reads
  draft_assembly:
    label: Assembly created from ONT nanopore reads.
    type: File
    outputSource: assembling/assembly
  polished_assembly:
    label: Assembly created from ONT nanopore reads, polished with Illumina reads.
    type: File
    outputSource: polishing/polished_assembly
  basecalling_reports:
    label: Reports generated by guppy basecallers.
    type: File[]
    outputSource: basecalling/sequencing_summary
  read_qc_reports:
    label: Reports regarding read quality.
    type: Directory[]
    outputSource: qc_reads/read_qc_directory
  assembly_qc_report:
    label: Report regarding quality of polished assembly.
    type: Directory
    outputSource: qc_assembly/report_directory

steps:
  raw_read_merging:
    label: Merge single-fast5-files into multi-fast5-files.
    run: ../cwl-tools/misc/mergeFast5Tool.cwl
    in:
      worker_threads: worker_threads
      fast5_directory: fast5_directory
    out: [merged_fast5_dir]
  basecalling:
    label: Basecall raw 1d2 nanopore reads.
    run: ../cwl-tools/basecalling/guppy1d2BasecallingWorkflow.cwl
    in:
      worker_threads: worker_threads
      fast5_directory: raw_read_merging/merged_fast5_dir
      flowcell_id: flowcell_id
      kit_id: kit_id
      config_name_1d: config_name_1d
      config_name_1d2: config_name_1d2
      enable_adapter_trimming:
        valueFrom: "true"
    out: [merged_reads, merged_1d_reads, merged_1d2_reads, sequencing_summary]
  assembling:
    label: Use miniasm + racon to create draft assembly from long reads.
    run: ../cwl-tools/assembly/miniasmX3Workflow.cwl
    in:
      worker_threads: worker_threads
      reads: basecalling/merged_reads
    out: [assembly, unitigs_layout]
  polishing:
    label: Use pilon to polish draft assembly with short reads.
    run: ../cwl-tools/polishing//pilonWorkflow.cwl
    in:
      worker_threads: worker_threads
      draft_genome: assembling/assembly
      reads_fwd: short_reads_fwd
      reads_rev: short_reads_rev
    out: [polished_assembly]
  qc_reads:
    label: Generate reports regarding read quality.
    run: ../cwl-tools/qc/readQCWorkflow.cwl
    scatter: reads
    in:
      worker_threads: worker_threads
      reads: [basecalling/merged_1d_reads, basecalling/merged_1d2_reads, short_reads_fwd, short_reads_rev]
    out: [read_qc_directory]
  qc_assembly:
    label: Generate reports regarding assembly quality.
    run: ../cwl-tools/qc/quastTool.cwl
    in:
      worker_threads: worker_threads
      assembly: polishing/polished_assembly
      reference_genome: reference_genome
    out: [report_directory]

s:author:
  - class: s:Person
    s:email: mailto:tom.tubb@googlemail.com
    s:name: Tom Tubbesing
s:dateCreated: "2019-06-26"
s:license: "https://spdx.org/licenses/GPL-3.0-or-later.html"

$namespaces:
  s: "http://schema.org/"
  edam: "http://edamontology.org/"
  cwltool: "http://commonwl.org/cwltool#"

$schemas:
  - https://schema.org/version/latest/schema.rdf
  - http://edamontology.org/EDAM_1.18.owl
